src/Middleware/HttpLogging/src/ResponseBufferingStream.cs

```diff
@@ -0,0 +1,446 @@
+            if ((sizeHint == 0 && remainingSize > 0) || (sizeHint > 0 && remainingSize >= sizeHint))
+            {
+                // We have capacity in the current segment
+#pragma warning disable CS8774 // Member must have a non-null value when exiting.
+                return;
+#pragma warning restore CS8774 // Member must have a non-null value when exiting.
+            }
+
+            AddSegment(sizeHint);
+        }
+
+        [MemberNotNull(nameof(_currentSegment))]
+        private void AddSegment(int sizeHint = 0)
+        {
+            if (_currentSegment != null)
+            {
+                // We're adding a segment to the list
+                if (_completedSegments == null)
+                {
+                    _completedSegments = new List<CompletedBuffer>();
+                }
+
+                // Position might be less than the segment length if there wasn't enough space to satisfy the sizeHint when
+                // GetMemory was called. In that case we'll take the current segment and call it "completed", but need to
+                // ignore any empty space in it.
+                _completedSegments.Add(new CompletedBuffer(_currentSegment, _position));
+            }
+
+            // Get a new buffer using the minimum segment size, unless the size hint is larger than a single segment.
+            _currentSegment = ArrayPool<byte>.Shared.Rent(Math.Max(_minimumSegmentSize, sizeHint));
+            _position = 0;
+        }
+
+        private async Task CopyToSlowAsync(Stream destination)
+        {
+            if (_completedSegments != null)
+            {
+                // Copy full segments
+                var count = _completedSegments.Count;
+                for (var i = 0; i < count; i++)
+                {
+                    var segment = _completedSegments[i];
+                    await destination.WriteAsync(segment.Buffer, 0, segment.Length);
+                }
+            }
+
+            if (_currentSegment is not null)
+            {
+                await destination.WriteAsync(_currentSegment, 0, _position);
+            }
+        }
+
+        public override void WriteByte(byte value)
+        {
+            if (_currentSegment != null && (uint)_position < (uint)_currentSegment.Length)
+            {
+                _currentSegment[_position] = value;
+            }
+            else
+            {
+                AddSegment();
+                _currentSegment[0] = value;
+            }
+
+            _position++;
+            _bytesWritten++;
+        }
+
+        // TODO inefficient, don't want to allocate array size of string,
+        // but issues with decoder APIs returning character count larger than
+        // required.
+        public string GetString(Encoding? encoding)
+        {
+            if (_currentSegment == null || encoding == null)
+            {
+                return "";
+            }
+
+            var result = new byte[_bytesWritten];
```

Comment:
I tried using Decoder APIs here but they seemed to not work with returning the right character count. So this is very inefficient at the moment.

Reply:
At a minimum we can rent this array

Reply:
I didn't look at the perf-characteristics, but
```c#
        public string GetString(Encoding? encoding)
        {
            if (_currentSegment == null || encoding == null)
            {
                return "";
            }

            var resultSize = 0;
            if (_completedSegments != null)
            {
                var count = _completedSegments.Count;
                for (var i = 0; i < count; ++i)
                {
                    var segment = _completedSegments[i];
                    resultSize += encoding.GetCharCount(segment.Span);
                }
            }
            resultSize += encoding.GetCharCount(_currentSegment.AsSpan(0, _position));

            return string.Create(resultSize, (this, encoding), static (buffer, state) =>
            {
                var (self, encoding) = state;
                var totalWritten = 0;
                var origBufferLength = buffer.Length;

                if (self._completedSegments != null)
                {
                    var count = self._completedSegments.Count;
                    for (var i = 0; i < count; ++i)
                    {
                        var segment = self._completedSegments[i];
                        var written = encoding.GetChars(segment.Span, buffer);
                        totalWritten += written;
                        buffer = buffer[written..];
                    }
                }

                totalWritten += encoding.GetChars(self._currentSegment.AsSpan(0, self._position), buffer);
                Debug.Assert(totalWritten == origBufferLength);
            });
        }
```
I don't like the double-iteration of the segements, but that's how these api work.
And the pro side it avoids some copying and allocating / renting a buffer for result.

From your code comment:

> issues with decoder APIs returning character count larger than required

Is this a bug?

Reply:
This wouldn't work, If a character is sliced between two spans, this will return the wrong character count. For example, this will fail on something like new string('あ', 10000), returning the wrong character count as for each chunk, it will add an extra character as the character is across two spans.

I thought the [Decoder](https://docs.microsoft.com/en-us/dotnet/api/system.text.decoder?view=net-5.0) apis would help here, but they ended up failing as well. I got one more idea for how to fix this that I'm going to try next.

Reply:
Ah, didn't think about that case.

> one more idea for how to fix this 

I'm interested to see this 👍🏻